File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateUseMemo.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  CompilerDiagnostic,
  CompilerError,
  ErrorCategory,
} from '../CompilerError';
import {
  FunctionExpression,
  HIRFunction,
  IdentifierId,
  SourceLocation,
} from '../HIR';
import {
  eachInstructionValueOperand,
  eachTerminalOperand,
} from '../HIR/visitors';
import {Result} from '../Utils/Result';

export function validateUseMemo(fn: HIRFunction): Result<void, CompilerError> {
  const errors = new CompilerError();
  const voidMemoErrors = new CompilerError();
  const useMemos = new Set<IdentifierId>();
  const react = new Set<IdentifierId>();
  const functions = new Map<IdentifierId, FunctionExpression>();
  const unusedUseMemos = new Map<IdentifierId, SourceLocation>();
  for (const [, block] of fn.body.blocks) {
    for (const {lvalue, value} of block.instructions) {
      if (unusedUseMemos.size !== 0) {
        /**
         * Most of the time useMemo results are referenced immediately. Don't bother
         * scanning instruction operands for useMemos unless there is an as-yet-unused
         * useMemo.
         */
        for (const operand of eachInstructionValueOperand(value)) {
          unusedUseMemos.delete(operand.identifier.id);
        }
      }
      switch (value.kind) {
        case 'LoadGlobal': {
          if (value.binding.name === 'useMemo') {
            useMemos.add(lvalue.identifier.id);
          } else if (value.binding.name === 'React') {
            react.add(lvalue.identifier.id);
          }
          break;
        }
        case 'PropertyLoad': {
          if (react.has(value.object.identifier.id)) {
            if (value.property === 'useMemo') {
              useMemos.add(lvalue.identifier.id);
            }
          }
          break;
        }
        case 'FunctionExpression': {
          functions.set(lvalue.identifier.id, value);
          break;
        }
        case 'MethodCall':
        case 'CallExpression': {
          // Is the function being called useMemo, with at least 1 argument?
          const callee =
            value.kind === 'CallExpression' ? value.callee : value.property;
          const isUseMemo = useMemos.has(callee.identifier.id);
          if (!isUseMemo || value.args.length === 0) {
            continue;
          }

          /*
           * If yes get the first argument and if it refers to a locally defined function
           * expression, validate the function
           */
          const [arg] = value.args;
          if (arg.kind !== 'Identifier') {
            continue;
          }
          const body = functions.get(arg.identifier.id);
          if (body === undefined) {
            continue;
          }

          if (body.loweredFunc.func.params.length > 0) {
            const firstParam = body.loweredFunc.func.params[0];
            const loc =
              firstParam.kind === 'Identifier'
                ? firstParam.loc
                : firstParam.place.loc;
            errors.pushDiagnostic(
              CompilerDiagnostic.create({
                category: ErrorCategory.UseMemo,
                reason: 'useMemo() callbacks may not accept parameters',
                description:
                  'useMemo() callbacks are called by React to cache calculations across re-renders. They should not take parameters. Instead, directly reference the props, state, or local variables needed for the computation',
                suggestions: null,
              }).withDetails({
                kind: 'error',
                loc,
                message: 'Callbacks with parameters are not supported',
              }),
            );
          }

          if (body.loweredFunc.func.async || body.loweredFunc.func.generator) {
            errors.pushDiagnostic(
              CompilerDiagnostic.create({
                category: ErrorCategory.UseMemo,
                reason:
                  'useMemo() callbacks may not be async or generator functions',
                description:
                  'useMemo() callbacks are called once and must synchronously return a value',
                suggestions: null,
              }).withDetails({
                kind: 'error',
                loc: body.loc,
                message: 'Async and generator functions are not supported',
              }),
            );
          }

          validateNoContextVariableAssignment(body.loweredFunc.func, errors);

          if (fn.env.config.validateNoVoidUseMemo) {
            if (!hasNonVoidReturn(body.loweredFunc.func)) {
              voidMemoErrors.pushDiagnostic(
                CompilerDiagnostic.create({
                  category: ErrorCategory.VoidUseMemo,
                  reason: 'useMemo() callbacks must return a value',
                  description: `This useMemo() callback doesn't return a value. useMemo() is for computing and caching values, not for arbitrary side effects`,
                  suggestions: null,
                }).withDetails({
                  kind: 'error',
                  loc: body.loc,
                  message: 'useMemo() callbacks must return a value',
                }),
              );
            } else {
              unusedUseMemos.set(lvalue.identifier.id, callee.loc);
            }
          }
          break;
        }
      }
    }
    if (unusedUseMemos.size !== 0) {
      for (const operand of eachTerminalOperand(block.terminal)) {
        unusedUseMemos.delete(operand.identifier.id);
      }
    }
  }
  if (unusedUseMemos.size !== 0) {
    /**
     * Basic check for unused memos, where the result of the call is never referenced. This runs
     * before DCE so it's more of an AST-level check that something, _anything_, cares about the value.
     *
     * This is easy to defeat with e.g. `const _ = useMemo(...)` but it at least gives us something to teach.
     * Even a DCE-based version could be bypassed with `noop(useMemo(...))`.
     */
    for (const loc of unusedUseMemos.values()) {
      voidMemoErrors.pushDiagnostic(
        CompilerDiagnostic.create({
          category: ErrorCategory.VoidUseMemo,
          reason: 'useMemo() result is unused',
          description: `This useMemo() value is unused. useMemo() is for computing and caching values, not for arbitrary side effects`,
          suggestions: null,
        }).withDetails({
          kind: 'error',
          loc,
          message: 'useMemo() result is unused',
        }),
      );
    }
  }
  fn.env.logErrors(voidMemoErrors.asResult());
  return errors.asResult();
}

function validateNoContextVariableAssignment(
  fn: HIRFunction,
  errors: CompilerError,
): void {
  for (const block of fn.body.blocks.values()) {
    for (const instr of block.instructions) {
      const value = instr.value;
      switch (value.kind) {
        case 'StoreContext': {
          errors.pushDiagnostic(
            CompilerDiagnostic.create({
              category: ErrorCategory.UseMemo,
              reason:
                'useMemo() callbacks may not reassign variables declared outside of the callback',
              description:
                'useMemo() callbacks must be pure functions and cannot reassign variables defined outside of the callback function',
              suggestions: null,
            }).withDetails({
              kind: 'error',
              loc: value.lvalue.place.loc,
              message: 'Cannot reassign variable',
            }),
          );
          break;
        }
      }
    }
  }
}

function hasNonVoidReturn(func: HIRFunction): boolean {
  for (const [, block] of func.body.blocks) {
    if (block.terminal.kind === 'return') {
      if (
        block.terminal.returnVariant === 'Explicit' ||
        block.terminal.returnVariant === 'Implicit'
      ) {
        return true;
      }
    }
  }
  return false;
}

File: compiler/crates/react_compiler_validation/src/validate_use_memo.rs
use std::collections::{HashMap, HashSet};

use react_compiler_diagnostics::{
    CompilerDiagnostic, CompilerDiagnosticDetail, CompilerError, ErrorCategory, SourceLocation,
};
use react_compiler_hir::{
    FunctionId, HirFunction, IdentifierId, InstructionValue,
    ParamPattern, PlaceOrSpread, Place, ReturnVariant, Terminal,
};
use react_compiler_hir::environment::Environment;
use react_compiler_hir::visitors::{
    each_instruction_value_operand_with_functions, each_terminal_operand,
};

/// Validates useMemo() usage patterns.
///
/// Port of ValidateUseMemo.ts.
/// Returns VoidUseMemo errors separately (for logging via logErrors, not as compile errors).
pub fn validate_use_memo(func: &HirFunction, env: &mut Environment) -> CompilerError {
    validate_use_memo_impl(func, &env.functions, &mut env.errors, env.config.validate_no_void_use_memo)
}

/// Information about a FunctionExpression needed for validation.
struct FuncExprInfo {
    func_id: FunctionId,
    loc: Option<SourceLocation>,
}

fn validate_use_memo_impl(
    func: &HirFunction,
    functions: &[HirFunction],
    errors: &mut CompilerError,
    validate_no_void_use_memo: bool,
) -> CompilerError {
    let mut void_memo_errors = CompilerError::new();
    let mut use_memos: HashSet<IdentifierId> = HashSet::new();
    let mut react_ids: HashSet<IdentifierId> = HashSet::new();
    let mut func_exprs: HashMap<IdentifierId, FuncExprInfo> = HashMap::new();
    let mut unused_use_memos: HashMap<IdentifierId, SourceLocation> = HashMap::new();

    for (_block_id, block) in &func.body.blocks {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            let lvalue = &instr.lvalue;
            let value = &instr.value;

            // Remove used operands from unused_use_memos
            if !unused_use_memos.is_empty() {
                for operand_id in each_instruction_value_operand_ids(value, functions) {
                    unused_use_memos.remove(&operand_id);
                }
            }

            match value {
                InstructionValue::LoadGlobal { binding, .. } => {
                    let name = binding.name();
                    if name == "useMemo" {
                        use_memos.insert(lvalue.identifier);
                    } else if name == "React" {
                        react_ids.insert(lvalue.identifier);
                    }
                }
                InstructionValue::PropertyLoad {
                    object, property, ..
                } => {
                    if react_ids.contains(&object.identifier) {
                        if let react_compiler_hir::PropertyLiteral::String(prop_name) = property {
                            if prop_name == "useMemo" {
                                use_memos.insert(lvalue.identifier);
                            }
                        }
                    }
                }
                InstructionValue::FunctionExpression { lowered_func, loc, .. } => {
                    func_exprs.insert(
                        lvalue.identifier,
                        FuncExprInfo {
                            func_id: lowered_func.func,
                            loc: *loc,
                        },
                    );
                }
                InstructionValue::CallExpression { callee, args, .. } => {
                    handle_possible_use_memo_call(
                        func,
                        functions,
                        errors,
                        &mut void_memo_errors,
                        &use_memos,
                        &func_exprs,
                        &mut unused_use_memos,
                        callee,
                        args,
                        lvalue,
                        validate_no_void_use_memo,
                    );
                }
                InstructionValue::MethodCall {
                    property, args, ..
                } => {
                    handle_possible_use_memo_call(
                        func,
                        functions,
                        errors,
                        &mut void_memo_errors,
                        &use_memos,
                        &func_exprs,
                        &mut unused_use_memos,
                        property,
                        args,
                        lvalue,
                        validate_no_void_use_memo,
                    );
                }
                _ => {}
            }
        }

        // Check terminal operands for unused_use_memos
        if !unused_use_memos.is_empty() {
            for operand_id in each_terminal_operand_ids(&block.terminal) {
                unused_use_memos.remove(&operand_id);
            }
        }
    }

    // Report unused useMemo results
    if !unused_use_memos.is_empty() {
        for loc in unused_use_memos.values() {
            void_memo_errors.push_diagnostic(
                CompilerDiagnostic::new(
                    ErrorCategory::VoidUseMemo,
                    "useMemo() result is unused",
                    Some(
                        "This useMemo() value is unused. useMemo() is for computing and caching values, not for arbitrary side effects"
                            .to_string(),
                    ),
                )
                .with_detail(CompilerDiagnosticDetail::Error {
                    loc: Some(*loc),
                    message: Some("useMemo() result is unused".to_string()),
                }),
            );
        }
    }

    void_memo_errors
}

#[allow(clippy::too_many_arguments)]
fn handle_possible_use_memo_call(
    _func: &HirFunction,
    functions: &[HirFunction],
    errors: &mut CompilerError,
    void_memo_errors: &mut CompilerError,
    use_memos: &HashSet<IdentifierId>,
    func_exprs: &HashMap<IdentifierId, FuncExprInfo>,
    unused_use_memos: &mut HashMap<IdentifierId, SourceLocation>,
    callee: &Place,
    args: &[PlaceOrSpread],
    lvalue: &Place,
    validate_no_void_use_memo: bool,
) {
    let is_use_memo = use_memos.contains(&callee.identifier);
    if !is_use_memo || args.is_empty() {
        return;
    }

    let first_arg = match &args[0] {
        PlaceOrSpread::Place(place) => place,
        PlaceOrSpread::Spread(_) => return,
    };

    let body_info = match func_exprs.get(&first_arg.identifier) {
        Some(info) => info,
        None => return,
    };

    let body_func = &functions[body_info.func_id.0 as usize];

    // Validate no parameters
    if !body_func.params.is_empty() {
        let first_param = &body_func.params[0];
        let loc = match first_param {
            ParamPattern::Place(place) => place.loc,
            ParamPattern::Spread(spread) => spread.place.loc,
        };
        errors.push_diagnostic(
            CompilerDiagnostic::new(
                ErrorCategory::UseMemo,
                "useMemo() callbacks may not accept parameters",
                Some(
                    "useMemo() callbacks are called by React to cache calculations across re-renders. They should not take parameters. Instead, directly reference the props, state, or local variables needed for the computation"
                        .to_string(),
                ),
            )
            .with_detail(CompilerDiagnosticDetail::Error {
                loc,
                message: Some("Callbacks with parameters are not supported".to_string()),
            }),
        );
    }

    // Validate not async or generator
    if body_func.is_async || body_func.generator {
        errors.push_diagnostic(
            CompilerDiagnostic::new(
                ErrorCategory::UseMemo,
                "useMemo() callbacks may not be async or generator functions",
                Some(
                    "useMemo() callbacks are called once and must synchronously return a value"
                        .to_string(),
                ),
            )
            .with_detail(CompilerDiagnosticDetail::Error {
                loc: body_info.loc,
                message: Some("Async and generator functions are not supported".to_string()),
            }),
        );
    }

    // Validate no context variable assignment
    validate_no_context_variable_assignment(body_func, functions, errors);

    if validate_no_void_use_memo && !has_non_void_return(body_func) {
        void_memo_errors.push_diagnostic(
            CompilerDiagnostic::new(
                ErrorCategory::VoidUseMemo,
                "useMemo() callbacks must return a value",
                Some(
                    "This useMemo() callback doesn't return a value. useMemo() is for computing and caching values, not for arbitrary side effects"
                        .to_string(),
                ),
            )
            .with_detail(CompilerDiagnosticDetail::Error {
                loc: body_info.loc,
                message: Some("useMemo() callbacks must return a value".to_string()),
            }),
        );
    } else if validate_no_void_use_memo {
        if let Some(callee_loc) = callee.loc {
            unused_use_memos.insert(lvalue.identifier, callee_loc);
        }
    }
}

fn validate_no_context_variable_assignment(
    func: &HirFunction,
    _functions: &[HirFunction],
    errors: &mut CompilerError,
) {
    let context: HashSet<IdentifierId> =
        func.context.iter().map(|place| place.identifier).collect();

    for (_block_id, block) in &func.body.blocks {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            if let InstructionValue::StoreContext { lvalue, .. } = &instr.value {
                if context.contains(&lvalue.place.identifier) {
                    errors.push_diagnostic(
                        CompilerDiagnostic::new(
                            ErrorCategory::UseMemo,
                            "useMemo() callbacks may not reassign variables declared outside of the callback",
                            Some(
                                "useMemo() callbacks must be pure functions and cannot reassign variables defined outside of the callback function"
                                    .to_string(),
                            ),
                        )
                        .with_detail(CompilerDiagnosticDetail::Error {
                            loc: lvalue.place.loc,
                            message: Some("Cannot reassign variable".to_string()),
                        }),
                    );
                }
            }
        }
    }
}

fn has_non_void_return(func: &HirFunction) -> bool {
    for (_block_id, block) in &func.body.blocks {
        if let Terminal::Return { return_variant, .. } = &block.terminal {
            if matches!(return_variant, ReturnVariant::Explicit | ReturnVariant::Implicit) {
                return true;
            }
        }
    }
    false
}

/// Collect all operand IdentifierIds from an InstructionValue.
/// Thin wrapper around canonical `each_instruction_value_operand_with_functions` that maps to ids.
fn each_instruction_value_operand_ids(
    value: &InstructionValue,
    functions: &[HirFunction],
) -> Vec<IdentifierId> {
    each_instruction_value_operand_with_functions(value, functions)
        .into_iter()
        .map(|p| p.identifier)
        .collect()
}

/// Collect all operand IdentifierIds from a Terminal.
/// Thin wrapper around canonical `each_terminal_operand` that maps to ids.
fn each_terminal_operand_ids(terminal: &Terminal) -> Vec<IdentifierId> {
    each_terminal_operand(terminal)
        .into_iter()
        .map(|p| p.identifier)
        .collect()
}
